### PR TITLE
WIP share a single watchman client

### PR DESCRIPTION
**Pull request by stefanpenner**
- [ ] handle error + restarting the client connection.
- [ ] more??
- [ ] test
- [ ] do i need to handle the lost connection case? Seems like fb-watchman may want to handle this.

### CODE REVIEW

File: src/watchman_watcher.js [hunk 1]
```diff
@@ -35,8 +37,24 @@ module.exports = WatchmanWatcher;
...
-  opts = common.assignOptions(this, opts);
+  common.assignOptions(this, opts);
+  this.id = ID++;
   this.root = path.resolve(dir);
+
+  var self = this;
+  this.onClientError = function(error) { self.emit('error', error); };
+  this.onHandleChangeEvent = this.handleChangeEvent.bind(this);
+  reconnecting = false;
+  this.onEnd = function() {
```
Comment (stefanpenner) [inline src/watchman_watcher.js]: WAT

File: src/watchman_watcher.js [hunk 2]
```diff
@@ -48,21 +66,28 @@ WatchmanWatcher.prototype.__proto__ = EventEmitter.prototype;
...
-  });
-  this.client.on('subscription', this.handleChangeEvent.bind(this));
-  this.client.on('end', function() {
-    console.warn('[sane] Warning: Lost connection to watchman, reconnecting..');
-    self.init();
-  });
+
+  if (!client) {
+    client = new watchman.Client();
+    client.setMaxListeners(100000);
```
Comment (stefanpenner) [inline src/watchman_watcher.js]: :(

File: src/watchman_watcher.js [hunk 3]
```diff
@@ -48,21 +64,28 @@ WatchmanWatcher.prototype.__proto__ = EventEmitter.prototype;
  * @private
  */
 
-WatchmanWatcher.prototype.init = function() {
-  if (this.client) {
-    this.client.removeAllListeners();
+WatchmanWatcher.prototype._detachOwnWatchmanHandlers = function() {
+  if (this.client && this.client.off) {
```
Comment (stefanpenner) [inline src/watchman_watcher.js]: WAT

File: src/watchman_watcher.js [hunk 4]
```diff
@@ -35,8 +37,22 @@ module.exports = WatchmanWatcher;
  */
 
 function WatchmanWatcher(dir, opts) {
-  opts = common.assignOptions(this, opts);
+  common.assignOptions(this, opts);
+  this.id = ID++;
   this.root = path.resolve(dir);
+
+  var self = this;
+  this.onClientError = function(error) { self.emit('error', error); };
```
Comment (stefanpenner) [inline src/watchman_watcher.js]: now errors from different clients may bubble up more then once (once per connected sane...?)

File: src/watchman_watcher.js [hunk 5]
```diff
@@ -184,7 +208,8 @@ WatchmanWatcher.prototype.init = function() {
  */
 
 WatchmanWatcher.prototype.handleChangeEvent = function(resp) {
-  assert.equal(resp.subscription, SUB_NAME, 'Invalid subscription event.');
+  if (resp.subscription !== SUB_NAME + this.id) { return; }
```
Comment (stefanpenner) [inline src/watchman_watcher.js]: make `SUB_NAME + this.id` a function or getter

**Comment by amasad** [comment]
Basic idea seems good. Implementation kinda ugly, maybe move the watchman client into it's own abstraction that handles reconnections, multiplexing etc.

**Comment by stefanpenner** [comment]
@amasad ya agreed, I think there are some issues with error handling/shutdown (from the watchman side). I'll try to find time to sort this out and get this in a better place.

**Comment by stefanpenner** [comment]
superseded by https://github.com/amasad/sane/pull/109

**Pull request closed without merging**